python/ray/data/_internal/execution/operators/output_splitter.py

```diff
@@ -0,0 +1,224 @@
+        # Invariant: buffer should always be large enough to equalize.
+        assert remainder >= 0, (remainder, buffer_size, allocation)
+
+        # Equally distribute remaining rows in buffer to outputs.
+        x = remainder // len(allocation)
+        allocation = [a + x for a in allocation]
+
+        # Execute the split.
+        for i, count in enumerate(allocation):
+            bundles = self._split_from_buffer(count)
+            for b in bundles:
+                b.output_split = i
+                self._output_queue.append(b)
+        self._buffer = []
+
+    def current_resource_usage(self) -> ExecutionResources:
+        return ExecutionResources(
+            object_store_memory=sum(b.size_bytes() for b in self._buffer)
+            + sum(b.size_bytes() for b in self._output_queue)
+        )
+
+    def progress_str(self) -> str:
+        if self._equal:
+            return f"{len(self._buffer)} buffered"
+        assert not self._buffer
+        return ""
+
+    def _dispatch_bundles(self) -> None:
+        # Dispatch all dispatchable bundles from the internal buffer.
+        # This may not dispatch all bundles when equal=True.
+        while self._buffer:
+            target_index = self._select_output_index()
+            target_bundle = self._pop_bundle_to_dispatch(target_index)
+            if self._can_safely_dispatch(target_index, target_bundle.num_rows()):
+                target_bundle.output_split = target_index
+                self._num_output[target_index] += target_bundle.num_rows()
+                self._output_queue.append(target_bundle)
+            else:
+                # Put it back and abort.
+                self._buffer.insert(0, target_bundle)
+                break
+
+    def _select_output_index(self) -> int:
+        # Greedily dispatch to the consumer with the least data so far.
+        i, _ = min(enumerate(self._num_output), key=lambda t: t[1])
+        return i
+
+    def _pop_bundle_to_dispatch(self, target_index: int) -> RefBundle:
+        # TODO implement locality aware bundle selection.
+        return self._buffer.pop(0)
+
+    def _can_safely_dispatch(self, target_index: int, nrow: int) -> bool:
+        if not self._equal:
+            # If not in equals mode, dispatch away with no buffer requirements.
+            return True
+        output_distribution = self._num_output.copy()
+        output_distribution[target_index] += nrow
+        buffer_requirement = self._calculate_buffer_requirement(output_distribution)
+        buffer_size = sum(b.num_rows() for b in self._buffer)
+        return buffer_size >= buffer_requirement
+
+    def _calculate_buffer_requirement(self, output_distribution: List[int]) -> int:
+        # Calculate the new number of rows that we'd need to equalize the row
+        # distribution after the bundle dispatch.
+        max_n = max(output_distribution)
+        return sum([max_n - n for n in output_distribution])
```

Comment:
It seems like this would suffer from the same issues as the consumer cursor route when it comes to large and/or skewed blocks, where we stall the consumer while waiting for the buffer to be filled with enough rows to ensure that we could rebalance in the future while we could be rebalancing now in order to unblock the consumers. I suppose that buffering in the operator further exacerbates this since we're not able to do a partial dispatch, as is possible in the consumer cursor route.

E.g. imagine you have 10 consumers specifying a 1024 `batch_size`, and the upstream dataset will eventually produce 10 blocks each with `1000 * 1024` rows, with one block produced every minute. When the first block arrives, we have 1024 batches read to be sent to consumers, but we haven't yet accumulated a block per consumer, so we have to stall. We'll ultimately stall for 10 minutes and then dispatch all blocks to all consumers, when we could have spent that 10 minutes pipelining with training.

Maybe we could see how this affects throughput on such a use case and explore a follow-up that splits based on some stalling mitigation heuristic, e.g. when there are `k * N * batch_size` rows in the buffer and we can't do the normal dispatch, do a bulk equalized split of the buffer (`N` is the # of consumers, `k` is some constant).

Reply:
We'll need to do buffering in the operator anyways for locality-aware dispatch. The larger the buffer, the better locality dispatch will be. See my other comment where I think we'll want to also enforce a minimum buffer size for locality reasons, in addition to correctness.

> We'll ultimately stall for 10 minutes and then dispatch all blocks to all consumers, when we could have spent that 10 minutes pipelining with training.

I believe that in real scenarios it's more likely to be a small factor op top of the initial pipeline startup. E.g., you wait for 1.5 blocks to be computed instead of 1.